### PR TITLE
Calendar component (2216)

### PR DESCRIPTION
**Pull request by Flaurite**
See https://github.com/jmix-framework/jmix/issues/2216

### CODE REVIEW

**Review by glebfox — CHANGES_REQUESTED** [review]

File: jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/component/data/LazyCalendarEventProvider.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ *     </li>
+ * </ul>
+ * By default {@link FullCalendar} calls {@link #onItemsFetch(ItemsFetchContext)} immediately after adding
+ * the provider. If calendar is not attached to the UI, loading will be performed after attaching component
+ * to the UI.
+ *
+ * @see LazyEntityCalendarEventRetriever
+ * @see CalendarEventRetriever
+ */
+public interface LazyCalendarEventProvider extends BaseCalendarEventProvider {
```
Comment (glebfox) [inline jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/component/data/LazyCalendarEventProvider.java:47]: Let's name it `CallbackCalendarEventProvider`, similar to `com.vaadin.flow.data.provider.CallbackDataProvider` that is used in `HasLazyDataView`

**Review by glebfox — CHANGES_REQUESTED** [review]

File: jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/FullCalendarFlowuiProperties.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.jmix.fullcalendarflowui;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties(prefix = "jmix.ui.fullcalendarflowui")
```
[the remark this answers is not in the document]
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/FullCalendarFlowuiProperties.java]: Fixed

File: jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/FullCalendarFlowuiProperties.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+public class FullCalendarFlowuiProperties {
+
+    String pickerDaysOfWeekEditShortcut;
```
[the remark this answers is not in the document]
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/FullCalendarFlowuiProperties.java]: Fixed

File: jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/sys/FullCalendarViewInstallDependencyInjector.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) Haulmont 2024. All Rights Reserved.
+ * Use is subject to license terms.
+ */
+
+package io.jmix.fullcalendarflowui.sys;
```
[the remark this answers is not in the document]
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/sys/FullCalendarViewInstallDependencyInjector.java:6]: Fixed

File: jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/component/FullCalendar.java [hunk 5]
```diff
@@ -0,0 +1,1518 @@
...
+import static io.jmix.fullcalendarflowui.kit.component.CalendarDateTimeUtils.*;
+import static io.jmix.fullcalendarflowui.kit.component.CalendarDateTimeUtils.parseAndTransform;
+
+/**
+ * UI component for visualizing events in a calendar using various views (month, week, etc.).
+ * <p>
+ * Component provides event rendering, drag-and-drop functionality, event editing, and customizable views.
+ */
+public class FullCalendar extends JmixFullCalendar implements ApplicationContextAware, InitializingBean {
+    private static final Logger log = LoggerFactory.getLogger(FullCalendar.class);
```
[the remark this answers is not in the document]
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/component/FullCalendar.java:58]: Fixed

File: jmix-fullcalendar/fullcalendar-flowui-kit/src/main/java/io/jmix/fullcalendarflowui/kit/component/CalendarDateTimeUtils.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        ZonedDateTime zonedDateTime = parseIsoDateTime(isoDateTime, zoneId);
+        return transformAsSystemDefault(zonedDateTime);
+    }
+
+    private static LocalDateTime transformAsSystemDefault(ZonedDateTime zonedDateTime) {
+        return zonedDateTime.withZoneSameInstant(ZoneId.systemDefault()).toLocalDateTime();
+    }
+
+    /**
+     * CAUTION! Copied from io.jmix.core.DateTimeTransformations#transformFromZdtInternal()
```
[the remark this answers is not in the document]
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar-flowui-kit/src/main/java/io/jmix/fullcalendarflowui/kit/component/CalendarDateTimeUtils.java]: Fixed

File: jmix-fullcalendar/fullcalendar-flowui-kit/src/main/java/io/jmix/fullcalendarflowui/kit/component/CalendarDateTimeUtils.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+     * <p>
+     * Transforms {@link ZonedDateTime} to the provided type.
+     *
+     * @param zonedDateTime date-time to transform
+     * @param javaType      java type to transform
+     * @return
```
[the remark this answers is not in the document]
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar-flowui-kit/src/main/java/io/jmix/fullcalendarflowui/kit/component/CalendarDateTimeUtils.java]: Fixed

File: jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/component/FullCalendar.java [hunk 8]
```diff
@@ -0,0 +1,1518 @@
...
+            addEventProviderInternal(eventProviderManager);
+        }
+    }
+
+    /**
+     * Adds new event provider.
+     *
+     * @param eventProvider event provider to add
+     */
+    public void addEventProvider(CalendarEventProvider eventProvider) {
```
[the remark this answers is not in the document]
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/component/FullCalendar.java]: Fixed

File: jmix-fullcalendar/fullcalendar/fullcalendar.gradle [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+apply plugin: 'io.jmix'
+
+
+group = 'io.jmix.fullcalendar'
```
[the remark this answers is not in the document]
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar/fullcalendar.gradle]: Fixed

File: jmix-fullcalendar/fullcalendar-flowui-kit/src/main/java/io/jmix/fullcalendarflowui/kit/component/model/CalendarView.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+import io.jmix.fullcalendarflowui.kit.component.JmixFullCalendar;
+
+/**
+ * Base interface for views in {@link JmixFullCalendar}.
+ *
+ * @see CalendarViewType
+ * @see <a href="https://fullcalendar.io/docs/view-object">FullCalendar docs :: view-object</a>
+ */
+public interface CalendarView extends HasEnumId<String> {
```
[the remark this answers is not in the document]
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar-flowui-kit/src/main/java/io/jmix/fullcalendarflowui/kit/component/model/CalendarView.java]: Fixed

File: jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/component/data/LazyCalendarEventProvider.java (hunk 1 above)
Comment (Flaurite) [reply jmix-fullcalendar/fullcalendar-flowui/src/main/java/io/jmix/fullcalendarflowui/component/data/LazyCalendarEventProvider.java]: I named it CallbackCalendar**Data**Provider to maintain consistency with other providers.

**Review by glebfox — APPROVED** [review]

**Pull request merged**